java/org/apache/catalina/connector/XForwardedRequest.java

```diff
@@ -0,0 +1,232 @@
+    }
+
+    @Override
+    public Enumeration<String> getHeaders(String name) {
+        Map.Entry<String,List<String>> header = getHeaderEntry(name);
+        if (header == null || header.getValue() == null) {
+            return Collections.enumeration(Collections.<String>emptyList());
+        }
+        return Collections.enumeration(header.getValue());
+    }
+
+    @Override
+    public int getIntHeader(String name) {
+        String value = getHeader(name);
+        if (value == null) {
+            return -1;
+        }
+        return Integer.parseInt(value);
+    }
+
+    @Override
+    public String getLocalName() {
+        return localName;
+    }
+
+    @Override
+    public int getLocalPort() {
+        return localPort;
+    }
+
+    @Override
+    public String getRemoteAddr() {
+        return this.remoteAddr;
+    }
+
+    @Override
+    public String getRemoteHost() {
+        return this.remoteHost;
+    }
+
+    @Override
+    public String getRequestId() {
+        if (this.requestId != null) {
+            return this.requestId;
+        }
+
+        return super.getRequest().getRequestId();
+    }
+
+    @Override
+    public String getScheme() {
+        return scheme;
+    }
+
+    @Override
+    public String getServerName() {
+        return serverName;
+    }
+
+    @Override
+    public int getServerPort() {
+        return serverPort;
+    }
+
+    public void removeHeader(String name) {
+        Map.Entry<String,List<String>> header = getHeaderEntry(name);
+        if (header != null) {
+            headers.remove(header.getKey());
+        }
+    }
+
+    public void setHeader(String name, String value) {
+        List<String> values = Collections.singletonList(value);
+        Map.Entry<String,List<String>> header = getHeaderEntry(name);
+        if (header == null) {
+            headers.put(name, values);
+        } else {
+            header.setValue(values);
+        }
+
+    }
+
+    public void setLocalName(String localName) {
+        this.localName = localName;
+    }
+
+    public void setLocalPort(int localPort) {
+        this.localPort = localPort;
+    }
+
+    public void setRemoteAddr(String remoteAddr) {
+        this.remoteAddr = remoteAddr;
+    }
+
+    public void setRemoteHost(String remoteHost) {
+        this.remoteHost = remoteHost;
+    }
+
+    public void setRequestId(String requestId) {
+        this.requestId = requestId;
+    }
+
+    public void setScheme(String scheme) {
+        this.scheme = scheme;
+    }
+
+    public void setSecure(boolean secure) {
+        super.getRequest().setAttribute(Globals.REMOTE_IP_FILTER_SECURE, Boolean.valueOf(secure));
+    }
+
+    public void setServerName(String serverName) {
+        this.serverName = serverName;
+    }
+
+    public void setServerPort(int serverPort) {
+        this.serverPort = serverPort;
+    }
+
+    @Override
+    public StringBuffer getRequestURL() {
+        return RequestUtil.getRequestURL(this);
+    }
+}
```

Comment:
I am a bit confused why this pull up is required here.

Reply:
The code for `XForwardedRequest` was cut and paste from the original, so TBH I'm not sure if it's needed

Reply:
OK, I see. But basically, it is an unrelated change?!

Reply:
@michael-o The extraction of `XForwardedRequest` from an internal class to a regular class is for reuse in the `RemoteIpValve`

Reply:
> @michael-o The extraction of `XForwardedRequest` from an internal class to a regular class is for reuse in the `RemoteIpValve`

Makes sense and now I see that those are two commits.

Reply:
@michael-o Yes, exactly.  I plan to add another commit here and add the enhancement to `RemoteIpValve`, and possibly to consolidate the duplicate code from `RemoteIpFilter` and `RemoteIpValve` into a single implementation for easier future maintenance.